core/chains/evm/gas/fee_history_estimator.go

```diff
@@ -0,0 +1,414 @@
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/services"
+	bigmath "github.com/smartcontractkit/chainlink-common/pkg/utils/big_math"
+
+	commonfee "github.com/smartcontractkit/chainlink/v2/common/fee"
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/gas/rollups"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+// metrics are thread safe
+var (
+	promFeeHistoryEstimatorGasPrice = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_price_updater",
+		Help: "Sets latest gas price (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promFeeHistoryEstimatorBaseFee = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "base_fee_updater",
+		Help: "Sets latest BaseFee (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promFeeHistoryEstimatorMaxPriorityFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_priority_fee_per_gas_updater",
+		Help: "Sets latest MaxPriorityFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promFeeHistoryEstimatorMaxFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_fee_per_gas_updater",
+		Help: "Sets latest MaxFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+)
+
+const (
+	MinimumBumpPercentage   = 10 // based on geth's spec
+	ConnectivityPercentile  = 85
+	BaseFeeBufferPercentage = 40
+)
+
+type FeeHistoryEstimatorConfig struct {
+	BumpPercent  uint16
+	CacheTimeout time.Duration
+	EIP1559      bool
+
+	BlockHistorySize uint64
+	RewardPercentile float64
+	HasMempool       bool
+}
+
+type feeHistoryEstimatorClient interface {
+	SuggestGasPrice(ctx context.Context) (*big.Int, error)
+	FeeHistory(ctx context.Context, blockCount uint64, rewardPercentiles []float64) (feeHistory *ethereum.FeeHistory, err error)
+}
+
+type FeeHistoryEstimator struct {
+	services.StateMachine
+
+	client  feeHistoryEstimatorClient
+	logger  logger.Logger
+	config  FeeHistoryEstimatorConfig
+	chainID *big.Int
+
+	gasPriceMu sync.RWMutex
+	gasPrice   *assets.Wei
+
+	dynamicPriceMu sync.RWMutex
+	dynamicPrice   DynamicFee
+
+	priorityFeeThresholdMu sync.RWMutex
+	priorityFeeThreshold   *assets.Wei
+
+	l1Oracle rollups.L1Oracle
+
+	wg     *sync.WaitGroup
+	stopCh services.StopChan
+}
+
+func NewFeeHistoryEstimator(lggr logger.Logger, client feeHistoryEstimatorClient, cfg FeeHistoryEstimatorConfig, chainID *big.Int, l1Oracle rollups.L1Oracle) *FeeHistoryEstimator {
+	return &FeeHistoryEstimator{
+		client:   client,
+		logger:   logger.Named(lggr, "FeeHistoryEstimator"),
+		config:   cfg,
+		chainID:  chainID,
+		l1Oracle: l1Oracle,
+		wg:       new(sync.WaitGroup),
+		stopCh:   make(chan struct{}),
+	}
+}
+
+func (u *FeeHistoryEstimator) Start(context.Context) error {
```

Comment:
nit: I think this got left over from when it was still called UniversalEstimator. Think we should update to `f` or `fh`
```suggestion
func (f *FeeHistoryEstimator) Start(context.Context) error {
```